src/backend/distributed/utils/attribute.c

```diff
@@ -0,0 +1,619 @@
+		time_t queryTime = time(0);
+		double cpu_time_used = ((double) (end - attributeToTenantStart)) / CLOCKS_PER_SEC;
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("attribute cpu counter (%f) to tenant: %s",
+									cpu_time_used, attributeToTenant)));
+		}
+
+		MultiTenantMonitor *monitor = GetMultiTenantMonitor();
+
+		LWLockAcquire(&monitor->lock, LW_SHARED);
+
+		monitor->periodStart = monitor->periodStart +
+							   ((queryTime - monitor->periodStart) /
+								CitusStatsTenantsPeriod) *
+							   CitusStatsTenantsPeriod;
+
+		int tenantIndex = FindTenantStats(monitor);
+
+		if (tenantIndex == -1)
+		{
+			tenantIndex = CreateTenantStats(monitor);
+		}
+		TenantStats *tenantStats = &monitor->tenants[tenantIndex];
+
+		LWLockAcquire(&tenantStats->lock, LW_EXCLUSIVE);
+
+		UpdatePeriodsIfNecessary(monitor, tenantStats);
+		tenantStats->lastQueryTime = queryTime;
+
+		ReduceScoreIfNecessary(monitor, tenantStats, queryTime);
+
+		/*
+		 * We do this after the reducing the scores so the scores in this period are not affected by the reduction.
+		 */
+		tenantStats->score += ONE_QUERY_SCORE;
+
+
+		/*
+		 * After updating the score we might need to change the rank of the tenant in the monitor
+		 */
+		while (tenantIndex != 0 &&
+			   monitor->tenants[tenantIndex - 1].score <
+			   monitor->tenants[tenantIndex].score)
+		{
+			LWLockAcquire(&monitor->tenants[tenantIndex - 1].lock, LW_EXCLUSIVE);
+
+			ReduceScoreIfNecessary(monitor, &monitor->tenants[tenantIndex - 1],
+								   queryTime);
+
+			TenantStats tempTenant = monitor->tenants[tenantIndex];
+			monitor->tenants[tenantIndex] = monitor->tenants[tenantIndex - 1];
+			monitor->tenants[tenantIndex - 1] = tempTenant;
+
+			LWLockRelease(&monitor->tenants[tenantIndex].lock);
+
+			tenantIndex--;
+		}
+		tenantStats = &monitor->tenants[tenantIndex];
+
+		if (attributeCommandType == CMD_SELECT)
+		{
+			tenantStats->selectCount++;
+			tenantStats->selectsInThisPeriod++;
+			tenantStats->totalSelectTime += cpu_time_used;
+		}
+		else if (attributeCommandType == CMD_INSERT)
+		{
+			tenantStats->insertCount++;
+			tenantStats->insertsInThisPeriod++;
+			tenantStats->totalInsertTime += cpu_time_used;
+		}
+
+		LWLockRelease(&monitor->lock);
+		LWLockRelease(&tenantStats->lock);
+
+		/*
+		 * We keep up to CitusStatsTenantsLimit * 3 tenants instead of CitusStatsTenantsLimit,
+		 * so we don't lose data immediately after a tenant is out of top CitusStatsTenantsLimit
+		 *
+		 * Every time tenant count hits CitusStatsTenantsLimit * 3, we reduce it back to CitusStatsTenantsLimit * 2.
+		 */
+		if (monitor->tenantCount >= CitusStatsTenantsLimit * 3)
+		{
+			LWLockAcquire(&monitor->lock, LW_EXCLUSIVE);
+			monitor->tenantCount = CitusStatsTenantsLimit * 2;
+			LWLockRelease(&monitor->lock);
+		}
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("total select count = %d, total CPU time = %f "
```

Comment:
NOTICE -> MultiTenantMonitoringLogLevel?

Reply:
Is this log really necessary? I think I can remove it